### PR TITLE
Mock Resampler in formula tests

### LINKED ISSUE #249: Some Formula tests are flaky
### What happened?

```
    async def test_3_phase_formulas(self, mocker: MockerFixture) -> None:
        """Test 3 phase formulas current formulas and their composition."""
        mockgrid = MockMicrogrid(grid_side_meter=False)
        mockgrid.add_batteries(3)
        mockgrid.add_ev_chargers(1)
        request_chan, channel_registry = await mockgrid.start(mocker)
        logical_meter = LogicalMeter(
            channel_registry,
            request_chan.new_sender(),
            microgrid.get().component_graph,
        )
        grid_current_recv = await logical_meter.grid_current()
        pool = EVChargerPool(channel_registry, request_chan.new_sender())
        ev_current_recv = await pool.total_current()
    
        engine = (grid_current_recv.clone() - ev_current_recv.clone()).build(
            "net_current"
        )
        net_current_recv = engine.new_receiver()
    
        await synchronize_receivers(
            [grid_current_recv, ev_current_recv, net_current_recv]
        )
    
        for _ in range(10):
            grid_amps = await grid_current_recv.receive()
            ev_amps = await ev_current_recv.receive()
            net_amps = await net_current_recv.receive()
    
>           assert grid_amps.value_p1 is not None and grid_amps.value_p1 > 0.0
E           assert (None is not None)
E            +  where None = Sample3Phase(timestamp=datetime.datetime(2023, 3, 10, 7, 9, 39, 78987, tzinfo=datetime.timezone.utc), value_p1=None, value_p2=None, value_p3=None).value_p1
```

### What did you expect instead?

We should fix them

### Affected version(s)

_No response_

### Affected part(s)

Unit, integration and performance tests (part:tests)

### Extra information

_No response_

### PR DESCRIPTION
**Pull request by sahas-subramanian-frequenz**
This allows most of the formula tests to be reliable.

Only the test for the `EVChargerPool.component_data` method still depends on a lower level mock, and it needs it because it combines resampled data with component state values.  If it is flaky, we'll have to add multi stage testing - with a fast test first, and a slower one if that fails, etc.

### CODE REVIEW

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py [hunk 1]
```diff
@@ -3,12 +3,20 @@
 
 """Formula generator for PV Power, from the component graph."""
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py:4]: > Fetch PV power from PV meters, rather than PV inverters

Can you please clarify/explain the reason of this change in the commit message?

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py [hunk 2]
```diff
@@ -66,3 +67,39 @@ def generate(self) -> FormulaEngine:
             builder.push_clipper(0.0, None)
 
         return builder.build()
+
+    def _get_pv_meters(self) -> abc.Set[int] | None:
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py]: Documentation is missing here.

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py [hunk 3]
```diff
@@ -66,3 +67,39 @@ def generate(self) -> FormulaEngine:
...
+        pv_inverters = list(
+            comp
+            for comp in component_graph.components()
+            if comp.category == ComponentCategory.INVERTER
+            and comp.type == InverterType.SOLAR
+        )
+        if not pv_inverters:
+            return None
+
+        pv_meters: set[int] = set()
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py]: I wonder if it'd be better to return this empty `set` rather than `None` when there is no PV inverters. If this comment applies, changes in the function definition will be needed.

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py [hunk 4]
```diff
@@ -110,3 +110,8 @@ def from_generator(
         ).generate()
```
Comment (daniel-zullo-frequenz) [inline src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:110]: I found the commit `Add methods for stopping running formula engines` is misleading because the changes in the commit are not only adding methods for stopping running formula engines but also updating existing ones.
So you might want to clarify the motivation or reason to update the existing methods or split those changes into a separate commit

File: tests/timeseries/mock_datapipeline.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            assert recv is not None
+            self._forward_tasks[request.get_channel_name()] = asyncio.create_task(
+                self._channel_forward_messages(
+                    recv,
+                    self._channel_registry.new_sender(request.get_channel_name()),
+                )
+            )
+
+    async def send_meter_power(self, values: list[float | None]) -> None:
+        """Send meter power."""
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/mock_datapipeline.py]: Probably is not worth it to add trivial documentation like this so I'm tempted to suggest to remove this kind of method documentation as there are other methods in the same file that don't add any documentation. If we keep it, then we should add documentation for the methods with missing docs

**Review by daniel-zullo-frequenz — COMMENTED** [review]
First review round done and I have a few suggestions to check for

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py (hunk 2 above)
Comment (shsms) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py]: It is a private function.  we don't need to document them unless they are not obvious.

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py (hunk 1 above)
Comment (shsms) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py:4]: done

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py (hunk 3 above)
Comment (shsms) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py]: true,  fixed.

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py (hunk 4 above)
Comment (shsms) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:110]: The updates to the other stop functions were just to call the formula engine stop functions, if they own formula engine instances.  So that when we call `stop` of a logical meter for example,  that calls the stop of the formula engines it owns as well.

I've updated the commit message.

File: tests/timeseries/mock_datapipeline.py (hunk 5 above)
Comment (shsms) [reply tests/timeseries/mock_datapipeline.py]: I guess I get lazy when writing tests.  I've improved the docs for these functions.

The docs can't be removed because they're public functions in tests.  Public test functions need to have a description, but not descriptions for args and return values.

File: src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py (hunk 2 above)
Comment (daniel-zullo-frequenz) [reply src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_pv_power_formula.py]: ok, I didn't know we can omit documentation for trivial internal methods

File: tests/timeseries/_formula_engine/test_formula_composition.py [hunk 6]
```diff
@@ -41,33 +42,56 @@ async def test_formula_composition(  # pylint: disable=too-many-locals
...
+        )  # -100 + -200
+        assert (
+            inv_calc_pow is not None
+            and inv_calc_pow.value is not None
+            and math.isclose(inv_calc_pow.value, -264.0)  # -300 + 36
+        )
+        assert (
+            main_pow is not None
+            and main_pow.value is not None
+            and math.isclose(main_pow.value, 100.0)
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/_formula_engine/test_formula_composition.py:79]: Alternatively we can use [pytest.approx](https://docs.pytest.org/en/7.1.x/reference/reference.html#id8) in tests. This is just an optional comment...

File: tests/timeseries/mock_microgrid.py [hunk 7]
```diff
@@ -48,16 +49,28 @@ class MockMicrogrid:  # pylint: disable=too-many-instance-attributes
...
-        self, grid_side_meter: bool, num_values: int = 2000, sample_rate_s: float = 0.01
+    mock_data: MockDataPipeline
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        grid_side_meter: bool,
+        api_client_streaming: bool = False,
+        num_values: int = 2000,
+        sample_rate_s: float = 0.01,
+        namespaces: int = 1,
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/mock_microgrid.py]: optionally I'd rename `namespaces` to `num_spaces` to better describe the variable and be consistent with other variables like `num_values`

File: tests/timeseries/mock_datapipeline.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    async def send_non_existing_component_value(self) -> None:
+        """Send a value for a non existing component."""
+        sample = Sample(self._next_ts, None)
+        await self._non_existing_component_sender.send(sample)
+
+    async def send_evc_current(self, values: list[list[float | None]]) -> None:
+        """Send the given values as resampler output for ev charger current."""
+        assert len(values) == len(self._ev_current_senders)
+        for chan, evc_values in zip(self._ev_current_senders, values):
+            assert len(evc_values) == 3
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/mock_datapipeline.py]: Optionally set a variable like `num_phases = 3` and compare it here

File: tests/timeseries/mock_datapipeline.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+    async def send_meter_power(self, values: list[float | None]) -> None:
+        """Send the given values as resampler output for meter power."""
+        assert len(values) == len(self._meter_power_senders)
+        for chan, value in zip(self._meter_power_senders, values):
+            sample = Sample(self._next_ts, value)
+            await chan.send(sample)
+
+    async def send_evc_power(self, values: list[float | None]) -> None:
+        """Send the given values as resampler output for ev charger power."""
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/mock_datapipeline.py]: s/ev/EV

File: tests/timeseries/mock_datapipeline.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            sample = Sample(self._next_ts, value)
+            await chan.send(sample)
+
+    async def send_non_existing_component_value(self) -> None:
+        """Send a value for a non existing component."""
+        sample = Sample(self._next_ts, None)
+        await self._non_existing_component_sender.send(sample)
+
+    async def send_evc_current(self, values: list[list[float | None]]) -> None:
+        """Send the given values as resampler output for ev charger current."""
```
Comment (daniel-zullo-frequenz) [inline tests/timeseries/mock_datapipeline.py]: s/ev/EV

**Review by daniel-zullo-frequenz — DISMISSED** [review]
I have only optional comments that you might want to check for. LGTM otherwise

File: tests/timeseries/mock_microgrid.py (hunk 7 above)
Comment (shsms) [reply tests/timeseries/mock_microgrid.py]: renamed to `num_namespaces`.

File: tests/timeseries/mock_datapipeline.py (hunk 8 above)
Comment (shsms) [reply tests/timeseries/mock_datapipeline.py]: It feels like this particular constant will lose some meaning if we give it a name.  I've added a comment instead for each occurrence instead.

File: tests/timeseries/mock_datapipeline.py (hunk 9 above)
Comment (shsms) [reply tests/timeseries/mock_datapipeline.py]: 🤦🏽 done

File: tests/timeseries/mock_datapipeline.py (hunk 10 above)
Comment (shsms) [reply tests/timeseries/mock_datapipeline.py]: done

**Review by daniel-zullo-frequenz — APPROVED** [review]
